Source/Core/Core/GeckoCodeConfig.cpp

```diff
@@ -2,17 +2,157 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include "Core/GeckoCodeConfig.h"
+
 #include <algorithm>
 #include <sstream>
 #include <string>
 #include <vector>
 
+#include <curl/curl.h>
+
 #include "Common/IniFile.h"
+#include "Common/Logging/Log.h"
 #include "Common/StringUtil.h"
-#include "Core/GeckoCodeConfig.h"
 
 namespace Gecko
 {
+static size_t DownloadCodesWriteCallback(void* contents, size_t size, size_t nmemb,
+                                         std::string* body)
+{
+  size_t realsize = size * nmemb;
+  body->insert(body->end(), reinterpret_cast<char*>(contents),
+               reinterpret_cast<char*>(contents) + realsize);
+  return realsize;
+}
+
+std::vector<GeckoCode> DownloadCodes(std::string gameid, bool* succeeded)
```

Comment:
Well, considering it's only ever used to build the endpoint string itself, you can just do:

```cpp
static std::string MakeEndpointURL(const std::string& game_id)
{
  constexpr const char* url_prefix = "http://geckocodes.org/txt.php?txt=";

  switch (game_id[0])
  {
  case 'R':
  case 'S':
  case 'G':
     return url_prefix + game_id;
  default:
    // All channels (WiiWare, VirtualConsole, etc) are identified by their first four characters
    return url_prefix + game_id.substr(0, 4);
}
```

then you just put the cURL stuff in its own function; now the lifetime of the cURL instance doesn't stick around through the whole parsing routine: 

```cpp
static std::pair<bool, std::string> RetrieveResponseBodyForGameID(const std::string& game_id)
{
  std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> curl{curl_easy_init(), curl_easy_cleanup};

  const std::string endpoint_url = MakeEndpointURL(game_id);
  curl_easy_setopt(curl, CURLOPT_URL, endpoint_url.c_str());
  curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5);

  std::string response_body;
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, DownloadCodesWriteCallback);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_body);

  CURLcode res = curl_easy_perform(curl);
  if (res != CURLE_OK)
  {
    ERROR_LOG(COMMON, "DownloadCodes: Curl error: %s", curl_easy_strerror(res));
    return std::make_pair(false, "");
  }

  long response_code = 0;
  curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
  if (response_code != 200)
  {
    WARN_LOG(COMMON, "DownloadCodes: Curl response code: %li", response_code);
    return std::make_pair(false, "");
  }

  return std::make_pair(true, response_code);
}
```

and then you just tie it together 

```cpp
std::vector<GeckoCode> DownloadCodes(const std::string& game_id, bool* succeeded)
{
  std::string response_body;
  std::tie(*succeeded, response_body) = RetrieveResponseBodyForGameID(game_id);

  if (!*succeeded)
    return {};

  /* Parse here */
}
```

and now you can make your `game_id` parameter a const reference like above instead of by value, and you also move all the variables that did the initial scaffolding completely out of the scope of the parsing routine. That's just my way of looking at it.